src/pd/client.rs

```diff
@@ -39,21 +36,58 @@ impl RpcClient {
                 .name_prefix(thd_name!(CLIENT_PREFIX))
                 .build(),
         );
-        for _ in 0..MAX_RETRY_TIMES {
-            match validate_endpoints(Arc::clone(&env), cfg, &security_mgr) {
-                Ok((client, members)) => {
-                    return Ok(RpcClient {
-                        cluster_id: members.get_header().get_cluster_id(),
-                        leader_client: LeaderClient::new(env, security_mgr, client, members),
-                    });
-                }
-                Err(e) => {
-                    warn!("validate PD endpoints failed"; "err" => ?e);
-                    thread::sleep(Duration::from_millis(RETRY_INTERVAL_MS));
+
+        if cfg.retry_interval.0.as_millis() == 0 {
+            let (client, members) = validate_endpoints(Arc::clone(&env), cfg, &security_mgr)?;
+
+            Ok(RpcClient {
+                cluster_id: members.get_header().get_cluster_id(),
+                leader_client: LeaderClient::new(env, security_mgr, client, members),
+            })
+        } else {
+            // We don't want to repeat the *same* error multiple times in `RETY_LOG_PER` count...
+            // But if the log message changes we need to output both.
+            // We can't compare PD Errors, but we can compare strings.
+            let mut cached_error = None;
+            let mut attempts_with_cached_error = 0;
+
+            while Some(attempts_with_cached_error) != cfg.retry_max_count {
+                match validate_endpoints(Arc::clone(&env), cfg, &security_mgr) {
+                    Ok((client, members)) => {
+                        return Ok(RpcClient {
+                            cluster_id: members.get_header().get_cluster_id(),
+                            leader_client: LeaderClient::new(env, security_mgr, client, members),
+                        });
+                    }
+                    Err(e) => {
+                        match cached_error.as_mut() {
+                            None => {
+                                warn!("validate PD endpoints failed"; "err" => ?e);
+                                attempts_with_cached_error += 1;
+                                cached_error = Some(format!("{}", e));
+                            }
+                            Some(ref mut cached) if **cached == format!("{}", e) => {
```

Comment:
Is `ref mut` necessary? It should be `&mut ` already.

Reply:
As per https://github.com/tikv/tikv/pull/4502#discussion_r286321003 this was altered.